File: aat-android/src/main/java/ch/bailu/aat/map/layer/MapPositionLayer.java
package ch.bailu.aat.map.layer;

import org.mapsforge.core.model.LatLong;

import ch.bailu.aat.preferences.map.SolidPositionLock;
import ch.bailu.aat_lib.coordinates.LatLongE6;
import ch.bailu.aat_lib.dispatcher.DispatcherInterface;
import ch.bailu.aat_lib.dispatcher.OnContentUpdatedInterface;
import ch.bailu.aat_lib.gpx.GpxInformation;
import ch.bailu.aat_lib.gpx.InfoID;
import ch.bailu.aat_lib.map.MapContext;
import ch.bailu.aat_lib.map.Point;
import ch.bailu.aat_lib.map.layer.MapLayerInterface;
import ch.bailu.aat_lib.preferences.StorageInterface;

public final class MapPositionLayer implements MapLayerInterface, OnContentUpdatedInterface {

    private final MapContext mcontext;

    public static final String LONGITUDE_SUFFIX ="longitude";
    public static final String LATITUDE_SUFFIX ="latitude";
    private static final String ZOOM_SUFFIX ="zoom";

    private final SolidPositionLock slock;

    private LatLong gpsLocation = new LatLong(0,0);

    private final StorageInterface storage;

    public MapPositionLayer(MapContext mc, StorageInterface storage, DispatcherInterface d) {
        mcontext = mc;
        this.storage = storage;

        slock = new SolidPositionLock(storage, mcontext.getSolidKey());

        loadState();
        d.addTarget(this, InfoID.LOCATION);
    }


    public void onMapCenterChanged(LatLong center) {
        if (gpsLocation.equals(center) == false) {
            disableLock();
        }
    }


    public void disableLock() {
        slock.setValue(false);
    }


    private void loadState() {
        gpsLocation = new LatLongE6(
                storage.readInteger(mcontext.getSolidKey() + LATITUDE_SUFFIX),
                storage.readInteger(mcontext.getSolidKey() + LONGITUDE_SUFFIX))
                .toLatLong();

        byte z = (byte) storage.readInteger(mcontext.getSolidKey() + ZOOM_SUFFIX);

        mcontext.getMapView().setZoomLevel(z);
        mcontext.getMapView().setCenter(gpsLocation);
    }


    private void setMapCenter() {
        if (slock.isEnabled()) {
            mcontext.getMapView().setCenter(gpsLocation);
        }
    }

    private void saveState() {
        LatLongE6 center = new LatLongE6(mcontext.getMapView().getMapViewPosition().getCenter());

        int zoom = mcontext.getMetrics().getZoomLevel();

        storage.writeInteger(mcontext.getSolidKey() + LATITUDE_SUFFIX, center.getLatitudeE6());
        storage.writeInteger(mcontext.getSolidKey() + LONGITUDE_SUFFIX, center.getLongitudeE6());
        storage.writeInteger(mcontext.getSolidKey() + ZOOM_SUFFIX, zoom);
    }

    @Override
    public void onLayout(boolean changed, int l, int t, int r, int b) {}


    @Override
    public void onPreferencesChanged(StorageInterface s, String key) {
        if (slock.hasKey(key)) {
            setMapCenter();
        }
    }



    @Override
    public void onAttached() {}

    @Override
    public void onDetached() {
        saveState();
    }


    @Override
    public void onContentUpdated(int iid, GpxInformation info) {
        gpsLocation = LatLongE6.toLatLong(info);
        setMapCenter();
    }


    @Override
    public void drawInside(MapContext mcontext) {}


    @Override
    public void drawForeground(MapContext mcontext) {}

    @Override
    public boolean onTap(Point tapPos) {
        return false;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/map/layer/MapPositionLayer.kt
package ch.bailu.aat_lib.map.layer

import ch.bailu.aat_lib.coordinates.LatLongE6.Companion.toLatLong
import ch.bailu.aat_lib.dispatcher.DispatcherInterface
import ch.bailu.aat_lib.dispatcher.TargetInterface
import ch.bailu.aat_lib.gpx.GpxInformation
import ch.bailu.aat_lib.gpx.InfoID
import ch.bailu.aat_lib.map.MapContext
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.preferences.location.SolidMapPosition
import ch.bailu.aat_lib.preferences.map.SolidPositionLock
import ch.bailu.aat_lib.util.Point
import org.mapsforge.core.model.LatLong

class MapPositionLayer(
    private val mcontext: MapContext,
    private val storage: StorageInterface,
    d: DispatcherInterface
) : MapLayerInterface, TargetInterface {
    private val slock: SolidPositionLock = SolidPositionLock(storage, mcontext.getSolidKey())
    private var gpsLocation = LatLong(0.0, 0.0)

    init {
        loadState()
        d.addTarget(this, InfoID.LOCATION)
    }

    fun onMapCenterChanged(center: LatLong) {
        if (gpsLocation != center) {
            disableLock()
        }
    }

    private fun disableLock() {
        slock.value = false
    }

    private fun loadState() {
        gpsLocation = SolidMapPosition.readPosition(storage, mcontext.getSolidKey()).toLatLong()
        val zoomLevel = SolidMapPosition.readZoomLevel(storage, mcontext.getSolidKey()).toByte()
        mcontext.getMapView().setZoomLevel(zoomLevel)
        mcontext.getMapView().setCenter(gpsLocation)
    }

    private fun setMapCenter() {
        if (slock.isEnabled) {
            mcontext.getMapView().setCenter(gpsLocation)
        }
    }

    private fun saveState() {
        val center = mcontext.getMapView().getMapViewPosition().center
        val zoomLevel = mcontext.getMapView().getMapViewPosition().zoomLevel.toInt()
        SolidMapPosition.writePosition(storage, mcontext.getSolidKey(), center)
        SolidMapPosition.writeZoomLevel(storage, mcontext.getSolidKey(), zoomLevel)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {}
    override fun onPreferencesChanged(storage: StorageInterface, key: String) {
        if (slock.hasKey(key)) {
            setMapCenter()
        }
    }

    override fun onAttached() {}
    override fun onDetached() {
        saveState()
    }

    override fun onContentUpdated(iid: Int, info: GpxInformation) {
        gpsLocation = toLatLong(info)
        setMapCenter()
    }

    override fun drawInside(mcontext: MapContext) {}
    override fun drawForeground(mcontext: MapContext) {}
    override fun onTap(tapPos: Point): Boolean {
        return false
    }
}
